### PR TITLE
Add initial transformer for node-lws

### PR DESCRIPTION
**Pull request by alexhultman**
This is the same as https://github.com/primus/primus/pull/439 but now things are published so tests should probably pass.

### CODE REVIEW

**Comment by lpinca** [comment]
Thank you!

Can you also add tests?
It should be sufficient to add a file called `lws.integrations.js` under `test` with the following content:
```js
'use strict';

require('./transformer.base')('lws');
```

**Comment by alexhultman** [comment]
Thanks for the comments. Did things update? I did an amend commit with the changes and force pushed it.

**Comment by lpinca** [comment]
Yes, it seems you have addressed everything.

**Comment by alexhultman** [comment]
Regarding the tests - wouldn't that be a responsibility of the node-lws module itself? I mean, this little (primus) wrapper is opt-in and in its separate, non-intrusive folder so it shouldn't break anything if not explicitly used, right?

**Comment by alexhultman** [comment]
Ah okay you only want a very very basic test that only does the require. Sure, but it will fail on Windows and in Node versions prior to 4.x so I would have to add some checks to only test it on supported platforms & versions right?

**Comment by lpinca** [comment]
We don't run tests on windows so you don't have to worry about that, but yes we should skip any unsupported node versions.

**Comment by lpinca** [comment]
We should also add some documentation where we document that the transformer will only run on Linux/OS X and on node >= 4, but we can do this later.

**Comment by alexhultman** [comment]
I did the same amend, force push.

**Comment by alexhultman** [comment]
Wow that is pretty cool. So you run a lot of tests using this - 8 failed and 604 passed. That's pretty decent for a first try.

**Comment by lpinca** [comment]
Yes we run integration tests in that way.
I can't look into the failing tests right now but hopefully I can find time later.
If you can investigate and see what is wrong, that would be great.

**Comment by jcrugzz** [comment]
@alexhultman keep up the great work!  Great to see more integrations here.

**Comment by alexhultman** [comment]
Yeah sure, thanks for the help. I'll run npm run integration till it works then.

**Comment by lpinca** [comment]
To speed up things a little (integration tests take a lot of time to run) you can run only lws tests by running
```
node_modules/.bin/mocha test/lws.integration.js
```
or write an npm script for it, but I'm probably saying something obvious here :)

**Comment by alexhultman** [comment]
No that's good to know. In case you didn't notice it yet - I'm not a Node.js/JS developer or anything like that so I'll take any hint you have for me :)

**Comment by alexhultman** [comment]
I'm sorry but I cannot even get the other tests to pass.

mocha test/websockets.integration.js (and any other transformer) just gives me 3 failing:

      1) "before each" hook: beforeEach for "exposes a compatible socket"
      2) "before each" hook: beforeEach for "exposes a compatible socket"
      3) "after each" hook: afterEach for "exposes a compatible socket"

I cannot find any info on this, any obvious miss?

**Comment by alexhultman** [comment]
Many of the fails (the fuck's) seems to come from the reconnect-after-end flaw though.

**Comment by lpinca** [comment]
This is weird, everything works on my machine.

**Comment by alexhultman** [comment]
Ah. Now it works. These tests are really nice!

**Comment by lpinca** [comment]
Almost all the fucks come from this [failure](https://github.com/primus/primus/blob/c11e35aa1bd23b463e7ff140738f978febbab2be/test/transformer.base.js#L1020).
This probably happens because the client does not receive this [message](https://github.com/primus/primus/blob/c11e35aa1bd23b463e7ff140738f978febbab2be/spark.js#L517).

The socket should first send the message, then close the connection. So in this case the order should be
```js
/*1.*/ service.send(socket, new Buffer(data), false);
/*2.*/ service.close(socket); 
```
I really have to work now :smile:

**Comment by alexhultman** [comment]
I've done a lot of debugging in the module itself. I didn't handle the server.close() at all but also there is a lot of small inconsistencies between ws and lws in terms of "am I supposed to trigger 'close' event after manually closing the socket, do I get close events for every socket when closing the server, etc etc. So I'm doing a lot of small tweaks to make the module more ws-like in behavior. The primus wrapper should be pretty much correct, I just need to tweak the module I think.

**Comment by lpinca** [comment]
This is a test case for my previous comment:

```js
'use strict';

const lws = require('lws');

const server = new lws.Server({ port: 3000 });

server.on('connection', socket => {
  console.log('connection');
  server.send(socket, new Buffer('foo'), false);
  server.close(socket);
});
```

```js
(function () {
  const ws = new WebSocket('ws://localhost:3000');

  ws.onopen = () => console.log('open');
  ws.onmessage = evt => console.log(evt.data);
  ws.onclose = () => console.log('close');
})();
```
The message never reaches the client. Tested on Firefox.

**Comment by alexhultman** [comment]
Ah! Nice catch. That will queue a sending, but close the socket before the actual sending starts. So I need to queue the closing until all messages has been sent (if there are messages in the queue).

**Comment by lpinca** [comment]
Almost all failures on travis are caused by that thing.

**Comment by alexhultman** [comment]
Wohoo! Only 2 failures now!

**Comment by alexhultman** [comment]
I get two "Uncaught Error: Digest method not supported" from this kind of call stack:

      at Error (native)
      at new Hash (crypto.js:50:18)
      at Object.Hash (crypto.js:49:12)
      at WebSocket.initAsClient (node_modules/ws/lib/WebSocket.js:596:23)
      at new WebSocket (node_modules/ws/lib/WebSocket.js:78:18)
      at Primus.opening (primus.js:2758:34)
      at Primus.reconnect (primus.js:2799:12)
      at Primus.reconnect (primus.js:1891:12)
      at Recovery.emitter (primus.js:143:12)
      at Recovery.emit (primus.js:238:35)
      at Recovery.delay (primus.js:629:14)
      at tickedtock (primus.js:935:14)

This doesn't seem to be something I can change, maybe it's some kind of missing things in some crypto.js thingy?

**Comment by lpinca** [comment]
Weird, are you running this in node?
No idea why https://github.com/websockets/ws/blob/40a9d686288b5d0be13f2bf2f3f5da07afc8cda2/lib/WebSocket.js#L596 throws the error.

**Comment by alexhultman** [comment]
I can get it to pass every test if I ignore to handle the server.close(). Am I supposed to send all messages before closing the server, just like when I close a single socket? Currently I just terminate the server in that server.close call.

**Comment by lpinca** [comment]
Yes I think so, If I remeber correctly we fire the `close` event on the transformer when we close the primus http server.

The default behaviour of node is to stop accepting new connections after calling `server.close()` but to keep active connections alive.

We close all [active connections](https://github.com/primus/primus/blob/c11e35aa1bd23b463e7ff140738f978febbab2be/index.js#L912-L914) just before calling `server.close()`.

I think all we need is to wait for all data to be sent (no new connections should come in).

**Comment by alexhultman** [comment]
Yeah I went with just ignoring it for now. It passes and it's far more important to support correct socket.close() than server.close() IMO. I will fix correct closing of the server later on. So I'll just publish 0.6.1 and change the package.json required version of lws so that it also reruns the tests and probably passes now.

**Comment by lpinca** [comment]
Sweet, good job!

**Comment by alexhultman** [comment]
WEEEE!

Thanks for the help!

**Comment by lpinca** [comment]
LGTM, but I would like @3rd-Eden to weigh in.

File: package.json [hunk 1]
```diff
@@ -90,7 +92,8 @@
     "sockjs": "0.3.x",
     "sockjs-client": "1.0.x",
     "through2": "2.0.x",
-    "ws": "1.0.x"
+    "ws": "1.0.x",
+    "lws": ">=0.6.1"
```
Comment (3rd-Eden) [inline package.json]: I would just use the same semver pattern as we are doing for the rest of the modules `0.6.x`

**Comment by 3rd-Eden** [comment]
For someone as opinionated as me, when it comes to code reviews.. I really have.. nothing to say here except for great job. Loving it! <3!

**Comment by 3rd-Eden** [comment]
@alexhultman It will be included in the 5.0 release.

**Pull request merged**